File: 2015/Day19/main.lua
local function parse_input(filename)
    local replacements = {}
    local melecule = ''

    for line in io.lines(filename) do
        if line:find('=>') then
            local from, to = line:match('(.+) => (.+)')
            table.insert(replacements, { from = from, to = to })
        elseif line ~= '' then
            melecule = line
        end
    end

    return replacements, melecule
end

local function mutate(molecule, replacements)
    local distinct_molecules = {}
    for _, rep in ipairs(replacements) do
        local start = 1
        while true do
            local i, j = molecule:find(rep.from, start, true)
            if not i then break end
            local new_molecule = molecule:sub(1, i - 1) .. rep.to .. molecule:sub(j + 1)
            distinct_molecules[new_molecule] = true
            start = i + 1
        end
    end
    local result = {}
    for k in pairs(distinct_molecules) do
        table.insert(result, k)
    end
    return result
end

local function shuffle_replacements(replacements)
    math.randomseed(os.time())
    for i = #replacements, 2, -1 do
        local j = math.random(i)
        replacements[i], replacements[j] = replacements[j], replacements[i]
    end
end

local function search(molecule, replacements)
    local target = molecule
    local mutations = 0

    while target ~= 'e' do
        local tmp = target
        for _, rep in ipairs(replacements) do
            local i, j = target:find(rep.to, 1, true)
            if i then
                target = target:sub(1, i - 1) .. rep.from .. target:sub(j + 1)
                mutations = mutations + 1
            end
        end

        if tmp == target then
            target = molecule
            mutations = 0
            shuffle_replacements(replacements)
        end
    end

    return mutations
end

local replacements, molecule = parse_input('input.txt')

print('Part 1: ' .. #mutate(molecule, replacements))
print('Part 2: ' .. search(molecule, replacements))

File: 2015/Day19/main.cpp
#include <algorithm>
#include <fstream>
#include <iostream>
#include <random>
#include <string>
#include <unordered_set>
#include <utility>
#include <vector>

struct Replacement {
  std::string from;
  std::string to;
  Replacement(std::string f, std::string t) : from(f), to(t) {}
};

std::pair<std::vector<Replacement>, std::string> parse_input(const std::string& filename) {
  std::ifstream file(filename);
  std::vector<Replacement> replacements;
  std::string molecule;
  std::string line;

  while (std::getline(file, line)) {
    if (line.find("=>") != std::string::npos) {
      size_t pos = line.find(" => ");
      replacements.push_back(Replacement(line.substr(0, pos), line.substr(pos + 4)));
    } else if (!line.empty()) {
      molecule = line;
    }
  }

  return {replacements, molecule};
}

std::unordered_set<std::string> mutate(const std::string& molecule, const std::vector<Replacement>& replacements) {
  std::unordered_set<std::string> distinct_molecules;
  for (const auto& rep : replacements) {
    size_t pos = 0;
    while ((pos = molecule.find(rep.from, pos)) != std::string::npos) {
      std::string new_molecule = molecule.substr(0, pos) + rep.to + molecule.substr(pos + rep.from.length());
      distinct_molecules.insert(new_molecule);
      pos += rep.from.length();
    }
  }
  return distinct_molecules;
}

int search(const std::string& molecule, std::vector<Replacement>& replacements) {
  std::string target = molecule;
  int mutations = 0;
  std::random_device rd;
  std::mt19937 gen(rd());

  while (target != "e") {
    std::string tmp = target;
    for (const auto& rep : replacements) {
      size_t index = target.find(rep.to);
      if (index != std::string::npos) {
        target = target.substr(0, index) + rep.from + target.substr(index + rep.to.length());
        mutations++;
        break;
      }
    }

    if (tmp == target) {
      target = molecule;
      mutations = 0;
      std::shuffle(replacements.begin(), replacements.end(), gen);
    }
  }

  return mutations;
}

int main() {
  auto [replacements, molecule] = parse_input("input.txt");

  std::cout << "Part 1: " << mutate(molecule, replacements).size() << std::endl;
  std::cout << "Part 2: " << search(molecule, replacements) << std::endl;

  return 0;
}
